sklearn/utils/_array_api.py

```diff
@@ -769,49 +792,66 @@ def _average(a, axis=None, weights=None, normalize=True, xp=None):
     return sum_ / scale
 
 
+def _xlogy(x, y, xp=None):
+    # TODO: Remove this once https://github.com/scipy/scipy/issues/21736 is fixed
+    xp, _, device_ = get_namespace_and_device(x, y, xp=xp)
+
+    with numpy.errstate(divide="ignore", invalid="ignore"):
+        temp = x * xp.log(y)
+    return xp.where(x == 0.0, xp.asarray(0.0, dtype=temp.dtype, device=device_), temp)
```

Comment:
Can't we simplify the arrays that we are creating in this function as well as in the `_nanmin, _nanmax, _nanmean`

Reply:
I tried, but passing an explicit `dtype` is mandatory to avoid upcasting to `float64` instead of keeping computation happening in `float32`. For instance, it would break `MaxAbsScaler` because the `.scale_` parameter would be `float64` when fit on `float32` data.

Reply:
I see. Yes that makes sense.